apps/remote_control/lib/lexical/remote_control/search/fuzzy.ex

```diff
@@ -0,0 +1,163 @@
+      path_to_refs: path_to_refs,
+      preprocessed_subjects: preprocessed_subjects
+    }
+  end
+
+  @spec match(t(), String.t()) :: [reference()]
+  def match(%__MODULE__{} = fuzzy, pattern) do
+    fuzzy.subject_to_refs
+    |> Stream.map(fn {subject, references} ->
+      case score(fuzzy, subject, pattern) do
+        {:ok, score} ->
+          {score, references}
+
+        :error ->
+          nil
+      end
+    end)
+    |> Stream.reject(&is_nil/1)
+    |> Enum.to_list()
+    |> List.keysort(0)
+    |> Enum.flat_map(&elem(&1, 1))
+  end
+
+  def add(%__MODULE__{} = fuzzy, %Entry{} = entry) do
+    updated_path_to_refs =
+      Map.update(fuzzy.path_to_refs, entry.path, [entry.ref], fn old_refs ->
+        [entry.ref | old_refs]
+      end)
+
+    string_subject = stringify(entry.subject)
+
+    updated_subject_to_refs =
+      Map.update(fuzzy.subject_to_refs, string_subject, [entry.ref], fn old_refs ->
+        [entry.ref | old_refs]
+      end)
+
+    updated_preprocessed_subjects =
+      Map.put_new_lazy(fuzzy.preprocessed_subjects, string_subject, fn ->
+        Scorer.preprocess(string_subject)
+      end)
+
+    %__MODULE__{
+      fuzzy
+      | path_to_refs: updated_path_to_refs,
+        subject_to_refs: updated_subject_to_refs,
+        preprocessed_subjects: updated_preprocessed_subjects
+    }
+  end
+
+  def has_path?(%__MODULE__{} = fuzzy, path) do
+    Map.has_key?(fuzzy.path_to_refs, path)
+  end
+
+  def has_subject?(%__MODULE__{} = fuzzy, subject) when is_binary(subject) do
+    Map.has_key?(fuzzy.subject_to_refs, subject)
+  end
+
+  def has_subject?(%__MODULE__{} = fuzzy, subject) do
+    has_subject?(fuzzy, inspect(subject))
+  end
+
+  def delete_path(%__MODULE__{} = fuzzy, path) do
+    refs = Map.get(fuzzy.path_to_refs, path, [])
+    fuzzy = drop_refs(fuzzy, refs)
+    %__MODULE__{fuzzy | path_to_refs: Map.delete(fuzzy.path_to_refs, path)}
+  end
+
+  def drop_refs(%__MODULE__{} = fuzzy, refs) do
+    ref_mapset = MapSet.new(refs)
+
+    reject_refs = fn {subject, refs} ->
+      {subject, Enum.reject(refs, &MapSet.member?(ref_mapset, &1))}
+    end
+
+    empty_refs? = fn
+      {_, []} -> true
+      {_, _} -> false
+    end
+
+    subject_to_refs =
+      fuzzy.subject_to_refs
+      |> Stream.map(reject_refs)
+      |> Stream.reject(empty_refs?)
+      |> Map.new()
+
+    all_subjects =
+      subject_to_refs
+      |> Map.keys()
+      |> MapSet.new()
+
+    path_to_refs =
+      fuzzy.path_to_refs
+      |> Stream.map(reject_refs)
+      |> Stream.reject(empty_refs?)
+      |> Map.new()
+
+    preprocessed_subjects =
+      fuzzy.preprocessed_subjects
+      |> Stream.filter(fn {subject, _} ->
+        MapSet.member?(all_subjects, subject)
+      end)
+      |> Map.new()
+
+    %__MODULE__{
+      fuzzy
+      | subject_to_refs: subject_to_refs,
+        path_to_refs: path_to_refs,
+        preprocessed_subjects: preprocessed_subjects
+    }
+  end
+
+  def update(%__MODULE__{} = fuzzy, entries) do
```

Comment:
I think this would be better off as another head of `add/2` or `add_list/2` if we want to keep them separate since there's really no difference in outcome between the two.

Reply:
add_all?

Reply:
I see you went with a guard. Works for me 👍